src/posix/platform/trel.cpp

```diff
@@ -0,0 +1,546 @@
+
+    return error;
+}
+
+static void ReceivePacket(int aSocket, otInstance *aInstance)
+{
+    struct sockaddr_in6 sockAddr;
+    socklen_t           sockAddrLen = sizeof(sockAddr);
+    ssize_t             ret;
+
+    memset(&sockAddr, 0, sizeof(sockAddr));
+
+    ret = recvfrom(aSocket, (char *)sRxPacketBuffer, sizeof(sRxPacketBuffer), 0, (struct sockaddr *)&sockAddr,
+                   &sockAddrLen);
+    VerifyOrDie(ret >= 0, OT_EXIT_ERROR_ERRNO);
+
+    sRxPacketLength = (uint16_t)(ret);
+
+    if (sRxPacketLength > sizeof(sRxPacketBuffer))
+    {
+        sRxPacketLength = sizeof(sRxPacketLength);
+    }
+
+    otLogDebgPlat("[trel] ReceivePacket() - received from [%s]:%d, id:%d, pkt:%s", Ip6AddrToString(&sockAddr.sin6_addr),
+                  ntohs(sockAddr.sin6_port), sockAddr.sin6_scope_id, BufferToString(sRxPacketBuffer, sRxPacketLength));
+
+    if (sEnabled)
+    {
+        otPlatTrelHandleReceived(aInstance, sRxPacketBuffer, sRxPacketLength);
+    }
+}
+
+static void InitPacketQueue(void)
+{
+    sTxPacketQueueTail = NULL;
+
+    // Chain all the packets in pool in the free linked list.
+    sFreeTxPacketHead = NULL;
+
+    for (uint16_t index = 0; index < OT_ARRAY_LENGTH(sTxPacketPool); index++)
+    {
+        TxPacket *packet = &sTxPacketPool[index];
+
+        packet->mNext     = sFreeTxPacketHead;
+        sFreeTxPacketHead = packet;
+    }
+}
+
+static void SendQueuedPackets(void)
+{
+    while (sTxPacketQueueTail != NULL)
+    {
+        TxPacket *packet = sTxPacketQueueTail->mNext; // tail->mNext is the head of the list.
+
+        if (SendPacket(packet->mBuffer, packet->mLength, &packet->mDestSockAddr) == OT_ERROR_INVALID_STATE)
+        {
+            otLogDebgPlat("[trel] SendQueuedPackets() - SendPacket() would block");
+            break;
+        }
+
+        // Remove the `packet` from the packet queue (circular
+        // linked list).
+
+        if (packet == sTxPacketQueueTail)
+        {
+            sTxPacketQueueTail = NULL;
+        }
+        else
+        {
+            sTxPacketQueueTail->mNext = packet->mNext;
+        }
+
+        // Add the `packet` to the free packet singly linked list.
+
+        packet->mNext     = sFreeTxPacketHead;
+        sFreeTxPacketHead = packet;
+    }
+}
+
+static void EnqueuePacket(const uint8_t *aBuffer, uint16_t aLength, const otSockAddr *aDestSockAddr)
+{
+    TxPacket *packet;
+
+    // Allocate an available packet entry (from the free packet list)
+    // and copy the packet content into it.
+
+    VerifyOrExit(sFreeTxPacketHead != NULL);
+    packet            = sFreeTxPacketHead;
+    sFreeTxPacketHead = sFreeTxPacketHead->mNext;
+
+    memcpy(packet->mBuffer, aBuffer, aLength);
+    packet->mLength       = aLength;
+    packet->mDestSockAddr = *aDestSockAddr;
+
+    // Add packet to the tail of TxPacketQueue circular linked-list.
+
+    if (sTxPacketQueueTail == NULL)
```

Comment:
nit: implement the circular linked list with a class since we are writing cpp?

Reply:
We can leave it for future. Again this is all existing code from ealier (not new code).
This file actually is all c style code (though it is named `trel.cpp` - maybe should rename to `trel.c`).

Reply:
SGTM.